File: Repo/opencog/cogutil/Perl/opencog-cogutil.pl
#!/usr/bin/env perl
# opencog-cogutil.pl
#
# OpenCog Cogutil - Perl Utility Library
# A collection of utility functions for OpenCog framework
#
# This single-file implementation demonstrates Perl's strengths:
# - Regular expressions for text processing
# - Hash tables (associative arrays) for data storage
# - References for complex data structures
# - CPAN-style modular design
# - Flexible syntax and TIMTOWTDI (There's More Than One Way To Do It)

use strict;
use warnings;
use v5.10;
use Time::HiRes qw(time);
use Data::Dumper;

# ===== Logger System =====
# Demonstrates: Blessed objects, hash references, method dispatch

package Logger {
    use strict;
    use warnings;
    
    my %LOG_LEVELS = (
        DEBUG => 0,
        INFO  => 1,
        WARN  => 2,
        ERROR => 3,
    );
    
    sub new {
        my ($class, $name, $level) = @_;
        $level //= 'INFO';
        
        my $self = {
            name  => $name,
            level => $level,
        };
        
        return bless $self, $class;
    }
    
    sub _current_timestamp {
        my @t = localtime(time);
        return sprintf("%02d:%02d:%02d", $t[2], $t[1], $t[0]);
    }
    
    sub _log {
        my ($self, $level, $message) = @_;
        
        if ($LOG_LEVELS{$level} >= $LOG_LEVELS{$self->{level}}) {
            my $timestamp = $self->_current_timestamp();
            say "[$timestamp] $level: $message";
        }
    }
    
    sub debug { $_[0]->_log('DEBUG', $_[1]) }
    sub info  { $_[0]->_log('INFO',  $_[1]) }
    sub warn  { $_[0]->_log('WARN',  $_[1]) }
    sub error { $_[0]->_log('ERROR', $_[1]) }
    
    sub set_level {
        my ($self, $level) = @_;
        $self->{level} = $level;
    }
}

# ===== Configuration Manager =====
# Demonstrates: Hash manipulation, file I/O, regex parsing

package Config {
    use strict;
    use warnings;
    
    sub new {
        my ($class) = @_;
        my $self = {
            data => {},
        };
        return bless $self, $class;
    }
    
    sub set {
        my ($self, $key, $value) = @_;
        $self->{data}{$key} = $value;
    }
    
    sub get {
        my ($self, $key, $default) = @_;
        $default //= '';
        return exists $self->{data}{$key} ? $self->{data}{$key} : $default;
    }
    
    sub has {
        my ($self, $key) = @_;
        return exists $self->{data}{$key};
    }
    
    sub load_from_file {
        my ($self, $filename) = @_;
        
        return 0 unless -e $filename;
        
        open(my $fh, '<', $filename) or return 0;
        
        while (my $line = <$fh>) {
            chomp $line;
            $line =~ s/^\s+|\s+$//g;  # Trim whitespace (Perl idiom)
            
            # Skip comments and empty lines
            next if $line =~ /^#/ || $line eq '';
            
            # Parse key=value pairs
            if ($line =~ /^([^=]+)=(.+)$/) {
                my ($key, $value) = ($1, $2);
                $key   =~ s/^\s+|\s+$//g;
                $value =~ s/^\s+|\s+$//g;
                $self->set($key, $value);
            }
        }
        
        close($fh);
        return 1;
    }
    
    sub dump {
        my ($self) = @_;
        for my $key (sort keys %{$self->{data}}) {
            say "$key = $self->{data}{$key}";
        }
    }
}

# ===== Timer Utility =====
# Demonstrates: High-resolution timing, closures

package Timer {
    use strict;
    use warnings;
    use Time::HiRes qw(time);
    
    sub new {
        my ($class, $name, $logger) = @_;
        my $self = {
            name       => $name,
            logger     => $logger,
            start_time => undef,
        };
        return bless $self, $class;
    }
    
    sub start {
        my ($self) = @_;
        $self->{start_time} = time();
    }
    
    sub stop {
        my ($self) = @_;
        return 0 unless defined $self->{start_time};
        
        my $elapsed = time() - $self->{start_time};
        if ($self->{logger}) {
            $self->{logger}->info(sprintf("%s completed in %.6f seconds", 
                                         $self->{name}, $elapsed));
        }
        return $elapsed;
    }
}

# ===== String Utilities =====
# Demonstrates: Regular expressions, array operations, map/grep

package StringUtils {
    use strict;
    use warnings;
    use Exporter 'import';
    our @EXPORT_OK = qw(str_split str_join to_lower to_upper trim 
                        camel_to_snake snake_to_camel reverse_str);
    
    sub str_split {
        my ($text, $delimiter) = @_;
        $delimiter //= ',';
        
        # Split and filter empty strings (using grep)
        return grep { $_ ne '' } 
               map { s/^\s+|\s+$//gr }  # Trim each element (non-destructive)
               split(quotemeta($delimiter), $text);
    }
    
    sub str_join {
        my ($strings, $delimiter) = @_;
        $delimiter //= ',';
        return join($delimiter, @$strings);
    }
    
    sub to_lower {
        my ($text) = @_;
        return lc($text);
    }
    
    sub to_upper {
        my ($text) = @_;
        return uc($text);
    }
    
    sub trim {
        my ($text) = @_;
        $text =~ s/^\s+|\s+$//g;
        return $text;
    }
    
    sub camel_to_snake {
        my ($text) = @_;
        $text =~ s/([A-Z])/_\L$1/g;  # Perl's case conversion in regex
        $text =~ s/^_//;
        return $text;
    }
    
    sub snake_to_camel {
        my ($text) = @_;
        $text =~ s/_([a-z])/\U$1/g;  # Uppercase after underscore
        return $text;
    }
    
    sub reverse_str {
        my ($text) = @_;
        return scalar reverse($text);  # Scalar context for string reverse
    }
}

# ===== Main Demonstration =====

package main;

use feature qw(say);

sub demonstrate_cogutil {
    say '=' x 70;
    say "OpenCog Cogutil - Perl Utility Library Demo";
    say "Showcasing Perl's strengths: Regex, hashes, TIMTOWTDI, text processing";
    say '=' x 70;
    say '';
    
    # Logger demonstration
    say "1. Logger Demonstration";
    say '-' x 50;
    my $logger = Logger->new("CogUtil", "INFO");
    $logger->info("Cogutil library initialized");
    $logger->debug("This debug message won't show (level too low)");
    $logger->warn("This is a warning message");
    $logger->error("This is an error message");
    
    $logger->set_level("DEBUG");
    $logger->debug("Now debug messages are visible");
    say '';
    
    # Config demonstration
    say "2. Configuration Manager";
    say '-' x 50;
    my $config = Config->new();
    $config->set("opencog.version", "1.0.0");
    $config->set("atomspace.enabled", "true");
    $config->set("cogserver.port", "17001");
    
    $logger->info("Configuration loaded:");
    $config->dump();
    say '';
    
    $logger->info("Port setting: " . $config->get("cogserver.port"));
    say '';
    
    # Timer demonstration
    say "3. Timer Utility";
    say '-' x 50;
    my $timer = Timer->new("Processing", $logger);
    $timer->start();
    $logger->info("Simulating some work...");
    
    # Simulate work using Perl idioms
    my $total = 0;
    $total += $_ for 1..1_000_000;
    
    $timer->stop();
    say '';
    
    # String utilities demonstration
    say "4. String Utilities with Regex";
    say '-' x 50;
    StringUtils->import(qw(str_split str_join to_lower to_upper trim 
                           camel_to_snake snake_to_camel reverse_str));
    
    $logger->info("String utilities demonstration:");
    my $text = "OpenCog,AtomSpace,CogServer,Cogutil";
    my @parts = str_split($text, ',');
    
    $logger->info("Split result:");
    say "  - $_" for @parts;
    
    my $joined = str_join(\@parts, " + ");
    $logger->info("Joined: $joined");
    
    $logger->info("Uppercase: " . to_upper("opencog rocks"));
    $logger->info("Lowercase: " . to_lower("OPENCOG ROCKS"));
    $logger->info("Trimmed: '" . trim("  spaced out  ") . "'");
    $logger->info("Reversed: " . reverse_str("Perl"));
    
    # Perl-specific: Case conversions with regex
    $logger->info("camelCase → snake_case: " . camel_to_snake("myVariableName"));
    $logger->info("snake_case → camelCase: " . snake_to_camel("my_variable_name"));
    say '';
    
    # Regular expression demonstration
    say "5. Regular Expression Power (Perl's Strength)";
    say '-' x 50;
    my $sample = "The AtomSpace uses Nodes and Links in OpenCog";
    
    # Pattern matching
    if ($sample =~ /AtomSpace/) {
        $logger->info("Found 'AtomSpace' in text");
    }
    
    # Capture groups
    if ($sample =~ /uses (\w+) and (\w+)/) {
        $logger->info("Captured: '$1' and '$2'");
    }
    
    # Substitution (non-destructive with /r modifier)
    my $modified = $sample =~ s/OpenCog/Perl/r;
    $logger->info("Substituted: $modified");
    
    # Word extraction (using match in list context)
    my @words = $sample =~ /\b([A-Z]\w+)\b/g;
    $logger->info("Capitalized words: " . join(", ", @words));
    say '';
    
    # Hash manipulation (Perl's associative arrays)
    say "6. Hash Operations (Associative Arrays)";
    say '-' x 50;
    my %atoms = (
        'Socrates' => 'human',
        'human'    => 'mortal',
        'mortal'   => 'being',
    );
    
    $logger->info("Hash contents:");
    say "  $_ => $atoms{$_}" for sort keys %atoms;
    
    # Hash slicing (Perl idiom)
    my @values = @atoms{qw(Socrates human)};
    $logger->info("Hash slice: " . join(", ", @values));
    say '';
    
    # Array and list operations
    say "7. Array Operations and List Processing";
    say '-' x 50;
    my @numbers = 1..10;
    
    # Map (transform)
    my @squares = map { $_ * $_ } @numbers;
    $logger->info("Squares: " . join(", ", @squares));
    
    # Grep (filter)
    my @evens = grep { $_ % 2 == 0 } @numbers;
    $logger->info("Even numbers: " . join(", ", @evens));
    
    # Reduce (using List::Util would be typical, but showing manual)
    my $sum = 0;
    $sum += $_ for @numbers;
    $logger->info("Sum: $sum");
    say '';
    
    # Context sensitivity (Perl's unique feature)
    say "8. Context Sensitivity (Perl Unique)";
    say '-' x 50;
    my @list = (1, 2, 3, 4, 5);
    
    # List context
    my @copy = @list;
    $logger->info("List context: " . join(", ", @copy));
    
    # Scalar context (gets length)
    my $count = @list;
    $logger->info("Scalar context (length): $count");
    say '';
    
    # Autovivification (Perl feature)
    say "9. Autovivification (Perl Magic)";
    say '-' x 50;
    my %deep;
    $deep{level1}{level2}{level3} = "value";  # Auto-creates intermediate hashes
    $logger->info("Deep hash value: " . $deep{level1}{level2}{level3});
    say '';
    
    $logger->info("Cogutil demonstration complete!");
    say '=' x 70;
    say "Perl strengths demonstrated:";
    say "  ✓ Powerful regular expressions (built into language)";
    say "  ✓ Hash tables (associative arrays) as first-class citizens";
    say "  ✓ Flexible syntax (TIMTOWTDI)";
    say "  ✓ Context sensitivity (list vs scalar)";
    say "  ✓ Text processing capabilities";
    say "  ✓ Autovivification for easy data structures";
    say "  ✓ Map/grep for functional programming";
    say '=' x 70;
}

# Run demonstration
demonstrate_cogutil();

File: Repo/opencog/cogutil/Lua/opencog-cogutil.lua
#!/usr/bin/env lua
--[[
opencog-cogutil.lua

OpenCog Cogutil - Lua Utility Library
A collection of utility functions and classes for OpenCog framework

This single-file implementation demonstrates Lua's strengths:
- Tables as universal data structure (arrays, dicts, objects)
- First-class functions and closures
- Metatables for object-oriented programming
- Lightweight and embeddable
- Simple and elegant syntax
]]

-- ===== Logger System =====
-- Demonstrates: Tables as objects, metatables for OOP

local LogLevel = {
    DEBUG = 1,
    INFO = 2,
    WARN = 3,
    ERROR = 4
}

local LogLevelNames = {
    [LogLevel.DEBUG] = "DEBUG",
    [LogLevel.INFO] = "INFO",
    [LogLevel.WARN] = "WARN",
    [LogLevel.ERROR] = "ERROR"
}

local Logger = {}
Logger.__index = Logger

function Logger:new(name, min_level)
    local self = setmetatable({}, Logger)
    self.name = name or "OpenCog"
    self.min_level = min_level or LogLevel.INFO
    return self
end

function Logger:log(level, message)
    if level >= self.min_level then
        local timestamp = os.date("%H:%M:%S")
        local level_name = LogLevelNames[level] or "UNKNOWN"
        print(string.format("[%s] %s: %s", timestamp, level_name, message))
    end
end

function Logger:debug(message)
    self:log(LogLevel.DEBUG, message)
end

function Logger:info(message)
    self:log(LogLevel.INFO, message)
end

function Logger:warn(message)
    self:log(LogLevel.WARN, message)
end

function Logger:error(message)
    self:log(LogLevel.ERROR, message)
end

function Logger:set_level(level)
    self.min_level = level
end

-- ===== Configuration Manager =====
-- Demonstrates: Tables as dictionaries, dynamic key access

local Config = {}
Config.__index = Config

function Config:new()
    local self = setmetatable({}, Config)
    self.settings = {}
    return self
end

function Config:set(key, value)
    self.settings[key] = value
end

function Config:get(key, default)
    return self.settings[key] or default
end

function Config:has(key)
    return self.settings[key] ~= nil
end

function Config:remove(key)
    self.settings[key] = nil
end

function Config:clear()
    self.settings = {}
end

function Config:dump()
    print("Configuration settings:")
    for key, value in pairs(self.settings) do
        print(string.format("  %s = %s", key, tostring(value)))
    end
end

function Config:size()
    local count = 0
    for _ in pairs(self.settings) do
        count = count + 1
    end
    return count
end

-- ===== Timer =====
-- Demonstrates: Closures, first-class functions

local Timer = {}
Timer.__index = Timer

function Timer:new(label, logger)
    local self = setmetatable({}, Timer)
    self.label = label or "Timer"
    self.logger = logger
    self.start_time = nil
    return self
end

function Timer:start()
    self.start_time = os.clock()
    if self.logger then
        self.logger:info(string.format("Timer '%s' started", self.label))
    end
end

function Timer:stop()
    if not self.start_time then
        error("Timer not started")
    end
    local elapsed = os.clock() - self.start_time
    if self.logger then
        self.logger:info(string.format("Timer '%s' stopped: %.4f seconds", self.label, elapsed))
    end
    return elapsed
end

function Timer:elapsed()
    if not self.start_time then
        return 0
    end
    return os.clock() - self.start_time
end

-- Higher-order function for timing
function measure_time(label, func)
    local start_time = os.clock()
    local result = func()
    local elapsed = os.clock() - start_time
    print(string.format("'%s' took %.4f seconds", label, elapsed))
    return result, elapsed
end

-- ===== String Utilities =====
-- Demonstrates: String manipulation, pattern matching, functional style

local StringUtils = {}

function StringUtils.split(str, delimiter)
    local result = {}
    local pattern = string.format("([^%s]+)", delimiter)
    for match in string.gmatch(str, pattern) do
        table.insert(result, match)
    end
    return result
end

function StringUtils.join(table, delimiter)
    return table.concat(table, delimiter)
end

function StringUtils.trim(str)
    return str:match("^%s*(.-)%s*$")
end

function StringUtils.to_upper(str)
    return string.upper(str)
end

function StringUtils.to_lower(str)
    return string.lower(str)
end

function StringUtils.starts_with(str, prefix)
    return string.sub(str, 1, #prefix) == prefix
end

function StringUtils.ends_with(str, suffix)
    return string.sub(str, -#suffix) == suffix
end

function StringUtils.contains(str, substring)
    return string.find(str, substring, 1, true) ~= nil
end

function StringUtils.reverse(str)
    return string.reverse(str)
end

function StringUtils.capitalize(str)
    return str:gsub("^%l", string.upper)
end

-- ===== Collection Utilities =====
-- Demonstrates: Functional programming with tables

local CollectionUtils = {}

function CollectionUtils.map(table, func)
    local result = {}
    for i, v in ipairs(table) do
        result[i] = func(v)
    end
    return result
end

function CollectionUtils.filter(table, predicate)
    local result = {}
    for _, v in ipairs(table) do
        if predicate(v) then
            table.insert(result, v)
        end
    end
    return result
end

function CollectionUtils.reduce(table, func, initial)
    local acc = initial
    for _, v in ipairs(table) do
        acc = func(acc, v)
    end
    return acc
end

function CollectionUtils.foreach(table, func)
    for i, v in ipairs(table) do
        func(v, i)
    end
end

function CollectionUtils.contains(table, value)
    for _, v in ipairs(table) do
        if v == value then
            return true
        end
    end
    return false
end

function CollectionUtils.size(table)
    local count = 0
    for _ in pairs(table) do
        count = count + 1
    end
    return count
end

-- ===== Memoization =====
-- Demonstrates: Closures, higher-order functions

function memoize(func)
    local cache = {}
    return function(...)
        local key = table.concat({...}, ",")
        if cache[key] == nil then
            cache[key] = func(...)
        end
        return cache[key]
    end
end

-- ===== Demonstration Function =====

local function demonstrate_cogutil()
    print("=" .. string.rep("=", 69))
    print("OpenCog Cogutil - Lua Implementation")
    print("=" .. string.rep("=", 69))
    print()
    
    -- Logger demonstration
    print("1. Logger System")
    print(string.rep("-", 50))
    local logger = Logger:new("OpenCog", LogLevel.INFO)
    
    logger:debug("This debug message won't show")
    logger:info("This is an info message")
    logger:warn("This is a warning message")
    logger:error("This is an error message")
    
    logger:set_level(LogLevel.DEBUG)
    logger:debug("Now debug messages are visible")
    print()
    
    -- Config demonstration
    print("2. Configuration Manager")
    print(string.rep("-", 50))
    local config = Config:new()
    config:set("opencog.version", "1.0.0")
    config:set("atomspace.enabled", "true")
    config:set("cogserver.port", "17001")
    
    logger:info("Configuration loaded:")
    config:dump()
    print()
    
    logger:info(string.format("Port setting: %s", config:get("cogserver.port")))
    logger:info(string.format("Config size: %d entries", config:size()))
    print()
    
    -- Timer demonstration
    print("3. Timer System")
    print(string.rep("-", 50))
    local timer = Timer:new("Processing", logger)
    timer:start()
    
    -- Simulate work
    local sum = 0
    for i = 1, 1000000 do
        sum = sum + i
    end
    
    timer:stop()
    print()
    
    -- Higher-order timing function
    print("4. Higher-Order Timing")
    print(string.rep("-", 50))
    local result, elapsed = measure_time("Computation", function()
        local product = 1
        for i = 1, 100 do
            product = product * 1.001
        end
        return product
    end)
    print()
    
    -- String utilities demonstration
    print("5. String Utilities")
    print(string.rep("-", 50))
    logger:info("String utilities demonstration:")
    local text = "OpenCog,AtomSpace,CogServer,Cogutil"
    local parts = StringUtils.split(text, ",")
    
    logger:info("Split result:")
    for _, part in ipairs(parts) do
        print(string.format("  - %s", part))
    end
    
    local joined = StringUtils.join(parts, " + ")
    logger:info(string.format("Joined: %s", joined))
    
    logger:info(string.format("Uppercase: %s", StringUtils.to_upper("opencog rocks")))
    logger:info(string.format("Lowercase: %s", StringUtils.to_lower("OPENCOG ROCKS")))
    logger:info(string.format("Trimmed: '%s'", StringUtils.trim("  spaced out  ")))
    logger:info(string.format("Capitalized: %s", StringUtils.capitalize("opencog")))
    logger:info(string.format("Reversed: %s", StringUtils.reverse("OpenCog")))
    print()
    
    -- Collection utilities demonstration
    print("6. Collection Utilities (Functional Style)")
    print(string.rep("-", 50))
    local numbers = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
    
    local squares = CollectionUtils.map(numbers, function(x) return x * x end)
    logger:info(string.format("Squares: %s", table.concat(squares, ", ")))
    
    local evens = CollectionUtils.filter(numbers, function(x) return x % 2 == 0 end)
    logger:info(string.format("Even numbers: %s", table.concat(evens, ", ")))
    
    local sum_result = CollectionUtils.reduce(numbers, function(acc, x) return acc + x end, 0)
    logger:info(string.format("Sum: %d", sum_result))
    print()
    
    -- Memoization demonstration
    print("7. Memoization")
    print(string.rep("-", 50))
    
    local function fibonacci(n)
        if n <= 1 then return n end
        return fibonacci(n-1) + fibonacci(n-2)
    end
    
    local fib_memo = memoize(fibonacci)
    
    local start = os.clock()
    local fib_result = fib_memo(30)
    local elapsed1 = os.clock() - start
    logger:info(string.format("Fibonacci(30) = %d (first call: %.4fs)", fib_result, elapsed1))
    
    start = os.clock()
    fib_result = fib_memo(30)
    local elapsed2 = os.clock() - start
    logger:info(string.format("Fibonacci(30) = %d (cached call: %.4fs)", fib_result, elapsed2))
    print()
    
    logger:info("Cogutil demonstration complete!")
    print("=" .. string.rep("=", 70))
    print("Lua strengths demonstrated:")
    print("  ✓ Tables as universal data structure")
    print("  ✓ Metatables for object-oriented programming")
    print("  ✓ First-class functions and closures")
    print("  ✓ Simple and elegant syntax")
    print("  ✓ Lightweight and fast")
    print("  ✓ Pattern matching for string manipulation")
    print("=" .. string.rep("=", 70))
end

-- Run demonstration if executed directly
if arg and arg[0]:match("opencog%-cogutil%.lua$") then
    demonstrate_cogutil()
end

-- Export module
return {
    Logger = Logger,
    LogLevel = LogLevel,
    Config = Config,
    Timer = Timer,
    StringUtils = StringUtils,
    CollectionUtils = CollectionUtils,
    measure_time = measure_time,
    memoize = memoize
}
